addons/point_of_sale/static/src/app/store/pos_store.js

```diff
@@ -661,75 +732,48 @@ export class PosStore extends Reactive {
     }
     async _loadMissingProducts(orders) {
         const missingProductIds = new Set([]);
+
         for (const order of orders) {
             for (const line of order.lines) {
                 const productId = line[2].product_id;
                 if (missingProductIds.has(productId)) {
                     continue;
                 }
-                if (!this.db.get_product_by_id(productId)) {
+                if (!this.models["product.product"].get(productId)) {
                     missingProductIds.add(productId);
                 }
             }
         }
-        if (!missingProductIds.size) {
-            return;
-        }
-        const products = await this.orm.call(
-            "pos.session",
-            "get_pos_ui_product_product_by_params",
-            [odoo.pos_session_id, { domain: [["id", "in", [...missingProductIds]]] }]
-        );
-        await this._loadMissingPricelistItems(products);
-        this._loadProductProduct(products);
+
+        await this.loadProducts([...missingProductIds]);
```

Comment:
when loading a new set of products (via a barcode scan), there will be `pricelist.item`s (linked to the loaded pricelists) that are not yet loaded. Load them as well and link those pricelist items to the products.

Reply:
https://github.com/odoo/odoo/blob/b02faa688234f8e7ccaac211678d3a23e5f3764c/addons/point_of_sale/models/pos_session.py#L2274

This method should be useful.